File: 340. Longest Substring with At Most K Distinct Characters/solution.py
class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        unique = {}
        l = 0
        i = 0
        j = 0
        while j < len(s):
            if s[j] not in unique:
                unique[s[j]] = 0
            unique[s[j]] += 1
            j += 1
            
            while len(unique) > k:
                unique[s[i]] -= 1
                if unique[s[i]] == 0:
                    unique.pop(s[i])
                    
                i += 1
                
            l = max(l, j - i)
            
        return l
                    

File: 340. Longest Substring with At Most K Distinct Characters/solution.java
class Solution {
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        Map<Character, Integer> charCountMap = new HashMap<>();
        int i = 0;
        int len = 0;
        Character character;
        Integer count;
        for (int j = 0; j < s.length(); j++) {
            // add char at j
            character = s.charAt(j);
            count = charCountMap.getOrDefault(character, 0);
            count++;
            charCountMap.put(character, count);
            
            // remove until capacity is created
            while (charCountMap.size() > k) {
                character = s.charAt(i);
                count = charCountMap.get(character);
                count--;
                if (count == 0) {
                    charCountMap.remove(character);
                } else {
                    charCountMap.put(character, count);
                }
                i++;
            }
            
            // calculate window length
            len = Math.max(len, j - i + 1);
        }
        return len;
    }
}
